pysindy/optimizers/trapping_sr3.py

```diff
@@ -0,0 +1,516 @@
+# import time
+import warnings
+
+import cvxpy as cp
+import numpy as np
+from scipy.linalg import cho_factor
+from scipy.linalg import cho_solve
+from sklearn.exceptions import ConvergenceWarning
+
+from ..utils import get_prox
+from ..utils import get_regularization
+from ..utils import reorder_constraints
+from .sr3 import SR3
+
+
+class trappingSR3(SR3):
+    """
+    Trapping variant of sparse relaxed regularized regression.
+
+    Attempts to minimize the objective function
+
+    .. math::
+
+        0.5\\|y-Xw\\|^2_2 + \\lambda \\times R(w)
+        + 0.5\\|Pw-A\\|^2_2/\\eta + \\delta_0(Cw-d)
+        + \\delta_{\\Lambda}(A)}
+
+    where :math:`R(w)` is a regularization function.
+    See the following references for more details:
+
+        Zheng, Peng, et al. "A unified framework for sparse relaxed
+        regularized regression: Sr3." IEEE Access 7 (2018): 1404-1423.
+
+        Champion, Kathleen, et al. "A unified sparse optimization framework
+        to learn parsimonious physics-informed models from data."
+        arXiv preprint arXiv:1906.10612 (2019).
+
+        New paper, Kaptanoglu et al. on trapping SINDy algorithms
+
+    Parameters
+    ----------
+    w_evo : bool, optional (default True)
+        If false, don't update w and just minimize over (m, A)
+
+    threshold : float, optional (default 0.1)
+        Determines the strength of the regularization. When the
+        regularization function R is the L0 norm, the regularization
+        is equivalent to performing hard thresholding, and lambda
+        is chosen to threshold at the value given by this parameter.
+        This is equivalent to choosing lambda = threshold^2 / (2 * nu).
+
+    eta : float, optional (default 1.0e20)
```

Comment:
Do the default values for `eta`, `alpha_A`, and `alpha_m` need to be this large?

Reply:
yes, in the sense that the default is to not use the stability terms. So the default eta >> 1 means that the algorithm should approximately reduce to the normal constrained SR3 that Kathleen wrote. This is better than starting with a "reasonable" value for eta, because this will vary for different datasets, and eta too small causes its own issues. The alpha_m and alpha_A values are proportional to a fraction eta (for solvability) so these are set very large too.

Reply:
Got it, thanks for the explanation!